File: python/leetcode/FlipMatrix.py
'''
Leetcode problem No 861 Score After Flipping Matrix
Solution written by Xuqiang Fang on 2 July, 2018
'''

class Solution(object):
    def matrixScore(self, A):
        """
        :type A: List[List[int]]
        :rtype: int
        """
        m = len(A); n = len(A[0])
        for i in range(m):
            if A[i][0] == 0:
                for j in range(n):
                    A[i][j] = 1 - A[i][j]
        for j in range(n):
            c = 0
            for i in range(m):
                if A[i][j]:
                    c += 1
            if c < (m+1) // 2:
                for x in range(m):
                    A[x][j] = 1 - A[x][j]
        s = 0
        for i in range(m):
            for j in range(n):
                if A[i][j]:
                    s += 2**(n-1-j)
        return s
                

def main():
    s = Solution()
    A = [[0,0,1,1],[1,0,1,0],[1,1,0,0]]
    print(s.matrixScore(A))
main()

File: java/leetcode/FlipMatrix.java
//Leetcode problem 861 Score After Flipping Matrix
//Solution written by Xuqiang Fang on 2 Junly, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class Solution{
    public int matrixScore(int[][] A) {
        int m = A.length;
        int n = A[0].length;
        for(int i=0; i<m; ++i){
            if(A[i][0] == 0){
                for(int j=0; j<n; ++j){
                    A[i][j] = 1 - A[i][j];
                }
            }
        }
        for(int j=0; j<n; ++j){
            int c = 0;
            for(int i=0; i<m; ++i){
                if(A[i][j] == 1){
                    c++;
                }
            }
            if(c < (m+1)/2){
                for(int x=0; x<m; ++x){
                    A[x][j] = 1 - A[x][j];
                }
            }
        }
        int s = 0;
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(A[i][j] == 1){
                    s += (int)Math.pow(2, n-1-j);
                }
            }
        }
        return s;
    }
}

public class FlipMatrix{
	public static void main(String[] args){
		Solution s = new Solution();
        int[][] A = {{0,0,1,1},{1,0,1,0},{1,1,0,0}};
        System.out.println(s.matrixScore(A));
	}
}
